RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cc

```diff
@@ -0,0 +1,460 @@
+//
+// Author: Felice Pantaleo, CERN
+//
+#include "CAHitQuadrupletGeneratorGPU.h"
+#include "RecoPixelVertexing/PixelTriplets/interface/ThirdHitPredictionFromCircle.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "TrackingTools/DetLayers/interface/BarrelDetLayer.h"
+
+#include "CellularAutomaton.h"
+
+#include "CommonTools/Utils/interface/DynArray.h"
+
+#include "FWCore/Utilities/interface/isFinite.h"
+
+#include <functional>
+
+namespace {
+
+template <typename T> T sqr(T x) { return x * x; }
+} // namespace
+
+using namespace std;
+
+constexpr unsigned int CAHitQuadrupletGeneratorGPU::minLayers;
+
+CAHitQuadrupletGeneratorGPU::CAHitQuadrupletGeneratorGPU(
+    const edm::ParameterSet &cfg,
+    edm::ConsumesCollector &iC)
+    : extraHitRPhitolerance(cfg.getParameter<double>(
+          "extraHitRPhitolerance")), // extra window in
+                                     // ThirdHitPredictionFromCircle range
+                                     // (divide by R to get phi)
+      maxChi2(cfg.getParameter<edm::ParameterSet>("maxChi2")),
+      fitFastCircle(cfg.getParameter<bool>("fitFastCircle")),
+      fitFastCircleChi2Cut(cfg.getParameter<bool>("fitFastCircleChi2Cut")),
+      useBendingCorrection(cfg.getParameter<bool>("useBendingCorrection")),
+      caThetaCut(cfg.getParameter<double>("CAThetaCut")),
+      caPhiCut(cfg.getParameter<double>("CAPhiCut")),
+      caHardPtCut(cfg.getParameter<double>("CAHardPtCut")) {
+  edm::ParameterSet comparitorPSet =
+      cfg.getParameter<edm::ParameterSet>("SeedComparitorPSet");
+  std::string comparitorName =
+      comparitorPSet.getParameter<std::string>("ComponentName");
+  if (comparitorName != "none") {
+    theComparitor.reset(SeedComparitorFactory::get()->create(
+        comparitorName, comparitorPSet, iC));
+  }
+  allocateOnGPU();
```

Comment:
Allocating the GPU memory from the EDProducer constructor does not work correctly with multiple devices as the current device is not set explicitly for the constructor. Please move the memory allocations to be done from `beginStreamGPU()`, where the current device is set according to the `edm::Stream` (which then set consistently for `acquireGPUCuda()` and `produceGPUCuda()`).

Personally I've so far used the pattern of constructing the GPU algorithm object in `beginStreamGPU()`. It won't work here directly though because of the `ConsumesCollector`. One option would be to abstract the configuration part (including the construction of `SeedComparitor`) to a separate class.